src/element-transformer/index.ts

```diff
@@ -0,0 +1,132 @@
+import * as ts from 'typescript';
+
+export default function transformer(program: ts.Program): ts.TransformerFactory<ts.SourceFile> {
+	return (context: ts.TransformationContext) => (file: ts.SourceFile) => visitNodeAndChildren(file, program, context);
+}
+
+function visitNodeAndChildren(f: ts.SourceFile, program: ts.Program, context: ts.TransformationContext): ts.SourceFile;
+function visitNodeAndChildren(node: ts.Node, program: ts.Program, context: ts.TransformationContext): ts.Node;
+function visitNodeAndChildren(node: ts.Node, program: ts.Program, context: ts.TransformationContext) {
+	return ts.visitEachChild(
+		decoratorVisit(node, program),
+		(childNode) => visitNodeAndChildren(childNode, program, context),
+		context
+	);
+}
+
+function decoratorVisit(node: ts.Node, program: ts.Program): ts.Node {
+	if (!(ts.isClassDeclaration(node) && node.decorators && node.name)) {
+		return node;
+	}
+	const decorator = node.decorators.find((dec) => {
+		return ts.isCallExpression(dec.expression) && dec.expression.expression.getText() === 'customElement';
+	});
+	if (!decorator) {
+		return node;
+	}
+
+	if (!ts.isCallExpression(decorator.expression) || decorator.expression.arguments.length !== 1) {
+		throw new Error('`customElement` decorator expects a single argument: string (tag name) or object');
+	}
+	let decArg = decorator.expression.arguments[0];
+	let didNormalize = false;
+
+	// normalize pass: transform @customElement('dojo-input') to @customElement({ tag: 'dojo-input' })
+	if (ts.isStringLiteral(decArg)) {
+		decArg = ts.createObjectLiteral([ts.createPropertyAssignment('tag', decArg)]);
+		didNormalize = true;
+	}
+
+	if (!ts.isObjectLiteralExpression(decArg)) {
+		throw new Error('`customElement` expects either a string (the tag name) or an object');
```

Comment:
again redundant, this would be a type error before it got here when the decorator is adjusted.